dwi_ml/training/trainers.py

```diff
@@ -1013,7 +1013,51 @@ def run_one_batch(self, data):
             Any other data returned when computing loss. Not used in the
             trainer, but could be useful anywhere else.
         """
-        raise NotImplementedError
+        # Data interpolation has not been done yet. GPU computations are done
+        # here in the main thread.
+        targets, ids_per_subj = data
+
+        # Dataloader always works on CPU. Sending to right device.
+        # (model is already moved).
+        targets = [s.to(self.device, non_blocking=True, dtype=torch.float)
+                   for s in targets]
+
+        # Getting the inputs points from the volumes.
+        # Uses the model's method, with the batch_loader's data.
+        # Possibly skipping the last point if not useful.
+        streamlines_f = targets
+        if isinstance(self.model, ModelWithDirectionGetter) and \
```

Comment:
You can delete all this if here. This only works on the basic model version.

But it could be an idea for you one day, to add a EOS in your data! We could move the EOS managemet in the basic model. To add to your todo list!